src/database/test/DatabaseTests.cpp

```diff
@@ -349,3 +355,430 @@ TEST_CASE("schema test", "[db]")
     auto av = db.getAppSchemaVersion();
     REQUIRE(dbv == av);
 }
+
+class SchemaUpgradeTestApplication : public TestApplication
+{
+  public:
+    using PreUpgradeFunc = std::function<void(SchemaUpgradeTestApplication&)>;
+
+  private:
+    PreUpgradeFunc const mPreUpgradeFunc;
+
+  public:
+    SchemaUpgradeTestApplication(VirtualClock& clock, Config const& cfg,
+                                 PreUpgradeFunc preUpgradeFunc)
+        : TestApplication(clock, cfg), mPreUpgradeFunc(preUpgradeFunc)
+    {
+    }
+
+    virtual void
+    actBeforeDBSchemaUpgrade() override
+    {
+        if (mPreUpgradeFunc)
+        {
+            try
+            {
+                mPreUpgradeFunc(*this);
+            }
+            catch (std::exception& e)
+            {
+                CLOG(FATAL, "Database")
+                    << __func__ << ": exception " << e.what()
+                    << " while calling pre-upgrade function";
+                throw;
+            }
+        }
+    }
+};
+
+// This tests upgrading from MIN_SCHEMA_VERSION to SCHEMA_VERSION by creating
+// ledger entries with the old MIN_SCHEMA_VERSION after database creation, then
+// validating their contents after the upgrade to SCHEMA_VERSION.
+TEST_CASE("schema upgrade test", "[db]")
+{
+    using OptLiabilities = stellar::optional<Liabilities>;
+    using AccOptLiabilitiesPair = std::pair<AccountEntry, OptLiabilities>;
```

Comment:
Good idea -- it was the way it was because in an earlier version of the code I was generating the ```AccountEntry``` and ```TrustLineEntry``` objects within the lambdas, and only passing in ```optional<Liabilities>```.  After I started generating the ```AccountEntry```/```TrustLineEntry``` objects in the caller of the lambas and passing them into the lambdas, it didn't occur to me that that could allow me to remove the ```optional<Liabilities>``` from the parameter lists.  I have now done so.